### ISSUE TITLE
Use `.graphqlconfig` File

### ISSUE THREAD

**Issue by 0xcaff** [jimkyndemeyer/js-graphql-intellij-plugin#125]
There's a de-facto standard for graphql metadata, [graphql-config](https://github.com/graphcool/graphql-config). Can this be used instead of the current format?

**Comment by jimkyndemeyer** [comment]
Hi Martin.

The config file format this plugin uses predates the graphql-config project, but I had discussions with the graphcool team when they began their work.

While I do see the value in converging on a single format, there are some important differences that makes adoption quite complex for this plugin.

The biggest difference is that this plugin relies on a single schema per IntelliJ/WebStorm project, whereas graphql-config allows multiple.

Like other language tooling on the IntelliJ platform, the project structure inside the IDE is what powers the various language features in this plugin. Again this difffers from graphql-config that uses glob patterns.

The bottom line is that adopting graphql-config cuts across all major features in this plugin. This includes schema resolution, completion, error highlighting, go to definition, and find usages. 

Realistically, I don't see where I could find the time to begin working on this in addition to maintaining the plugin. Right now #121 is the highest priority to ensure that the plugin keeps working as expected as the GraphQL spec evolves.

/Jim

**Comment by RIP21** [comment]
@jimkyndemeyer hey. Is there any help I can give? If you will guide me a bit where to watch I may be able to help out with adopting `.graphqlconfig`config format.

**Comment by jimkyndemeyer** [comment]
Hi Andrii.

Thanks for offering to help.

Given the scale of the change, I don't think it's realistic to work on this unless you are very experienced with the IntellIJ Platform SDK.

But, I've been thinking about an interim solution that can sync compatible properties from a `.graphqlconfig` file into the `graphql.config.json` file. While this doesn't cover all potential use cases and variants of a `.graphlconfig` setup, it could help developers with keeping the files in sync.

See `JSGraphQLConfigurationProvider#contentsChanged` for how the plugin updates the UI when the list of endpoints changes. If should be possible to also respond to changes in a `.graphlconfig` file and write compatible properties to the `graphql.config.json` file, e.g. the file or url that contains the schema.

Let me know if you have any questions.

/Jim

**Comment by RIP21** [comment]
@jimkyndemeyer I don't have any experience with it TBH :) 
Also, if not a config. which is a minor thing TBH, what about support for the `.schema` files? Cause now all the types must be described using `schema.json` which has a completely different format, but without it no syntax highlight and IntelliSense support. Maybe I'm doing something wrong, but it's the biggest issue I have with this plugin, that's why I don't use it at all (cause highlighting without types are all red)

**Comment by jimkyndemeyer** [comment]
Version 1.6.0 of the plugin introduced support for .graphql and .graphqls files that describe the schema using the Schema Definition Language.

https://github.com/jimkyndemeyer/js-graphql-intellij-plugin/blob/62c3dc1b632e93cd4d4c56b0395e7fb12a0ac6c9/resources/META-INF/graphql.config.json#L6-L7

This was added to support Relay Modern projects. It that what you're looking for?

**Comment by RIP21** [comment]
@jimkyndemeyer exactly! Missed that info about the fact that I need to rename it :) (better make it multiline instead :D)
Thanks a lot! (btw you can delete all unrelated comments)

**Comment by jimkyndemeyer** [comment]
@RIP21 Glad it worked out. I've deleted our comments that were somewhat unrelated to this issue.

**Comment by ForsakenHarmony** [comment]
So it won't ever be possible to have multiple schemas?

I have a project with 2 directories that have different schemas, but don't want to have to open webstorm twice

**Comment by jimkyndemeyer** [comment]
The 2.0 alpha makes it possible to have multiple schemas in a single project using IntelliJ/WebStorm "Scopes". You can configure them for the project in the IDE Settings dialog.

With that in place it should become possible to automatically sync the include/exclude rules in a watched graphql-config file such that the plugin language features respect said rules.

See [2.0.0-alpha-2](https://github.com/jimkyndemeyer/js-graphql-intellij-plugin/releases/tag/2.0.0-alpha-2) if you'd like to try it and help test it.

/cc @schickling

**Comment by fetis** [comment]
it's quite strange but the plugin is listed here as supported this format https://github.com/prismagraphql/graphql-config

**Comment by jimkyndemeyer** [comment]
@fetis Note the "(pending)" at the end of the sentence.

**Comment by fetis** [comment]
omg. how reads them all? :)

**Comment by jimkyndemeyer** [comment]
Released in https://github.com/jimkyndemeyer/js-graphql-intellij-plugin/releases/tag/2.0.0-alpha-3

**Comment by ForsakenHarmony** [comment]
would you consider moving to https://github.com/graphql/graphql-language-service/ as well?

**Comment by jimkyndemeyer** [comment]
I have, and I've been in talks with both Prisma and Facebook engineers.

It really comes down to trade-offs. While graphql-language-service does offer some advantages from shared tooling, it has the following drawbacks:

- It's JavaScript-based and requires Node.js tooling on the IntelliJ platform. This excludes Android Studio and Intellij Community users.
- Integration with the more powerful features in the IntelliJ platform is much more complicated when calling out to an external language service.

What really made the difference for v2 is the great work being done on graphql-java, which enables this plugin to get features such as validation and GraphQL SDL tools directly from the Java ecosystem.

**Comment by ForsakenHarmony** [comment]
oh, so you're moving away from the other javascript language server as well?

**Comment by 0xcaff** [comment]
I'm not sure how graphql-language-service is structured, but would it be possible to wrap graphql-language-service using the nashorn JavaScript engine to remove the dependency of a nodejs environment.

**Comment by ForsakenHarmony** [comment]
I doubt it supports running stuff written for node.js

**Comment by jimkyndemeyer** [comment]
Closing as v2 alpha now uses .graphqlconfig files and can migrate the legacy graphql.config.json files automatically.
